### PR TITLE
Added ssmManagedInstances plugin and spec

### PR DESCRIPTION
**Pull request by m-nouman-suleman**
## Task
- https://scalock.atlassian.net/browse/SAAS-3375

### CODE REVIEW

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+// TODO: MOVE TO EC2
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'SSM Managed Instances',
+    category: 'SSM',
+    domain: 'Identity Access and Management',
+    description: 'Ensure that all Amazon EC2 instances are managed by AWS Systems Manager (SSM)',
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:9]:
```suggestion
    description: 'Ensure that all Amazon EC2 instances are managed by AWS Systems Manager (SSM).',
```

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    more_info: 'Systems Manager simplifies AWS cloud resource management, shortens the time to detect and resolve operational problems, and makes it easy to operate and manage your instances securely at scale.',
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:10]: Rephrase

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    recommended_action: 'Configure AWS EC2 instance as managed instances',
+    link: 'https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-paramstore-about.html#sysman-paramstore-securestring',
+    apis: ['EC2:describeInstances', 'SSM:describeInstanceInformation', 'STS:getCallerIdentity'],
+    settings: {},
+
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:15]:
```suggestion

```

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            if (!describeInstances) return rcb();
+
+            if (!describeInstances.data || describeInstances.err) {
+                helpers.addResult(results, 3, 'Unable to query for EC2 instances: ' + helpers.addError(describeInstances), region);
+                return rcb();
+            } else if (!describeInstances.data.length) {
+                helpers.addResult(results, 0, 'No EC2 instances found', region);
+                return rcb();
+            }
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:37]:
```suggestion
            }

            if (!describeInstances.data.length) {
                helpers.addResult(results, 0, 'No EC2 instances found', region);
                return rcb();
            }
```

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            if (!describeInstanceInformation) return rcb();
+
+            if (describeInstanceInformation.err || !describeInstanceInformation.data) {
+                helpers.addResult(results, 3,
+                    'Unable to query for SSM instance information: ' + helpers.addError(describeInstanceInformation), region);
+                return rcb();
+            } else if (!describeInstanceInformation.data.length) {
+                helpers.addResult(results, 2, 'No EC2 instance is managed by AWS Systems Manager', region);
+                return rcb();
+            }
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:51]:
```suggestion

            if (!describeInstanceInformation || describeInstanceInformation.err || !describeInstanceInformation.data) {
                helpers.addResult(results, 3,
                    'Unable to query for SSM instance information: ' + helpers.addError(describeInstanceInformation), region);
                return rcb();
            }
```

File: plugins/aws/ssm/ssmManagedInstances.spec.js [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+                            "VolumeId": "vol-0ebea24b6b5ab89d5"
+                        }
+                    }
+                ],
+                "ClientToken": "",
+                "EbsOptimized": false,
+                "EnaSupport": true,
+                "Hypervisor": "xen",
+                "IamInstanceProfile": {
+                    "Arn": "arn:aws:iam::101363889637:instance-profile/AmazonSSMRoleForInstancesQuickSetup",
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.spec.js:52]:
```suggestion
                    "Arn": "arn:aws:iam::000011112222:instance-profile/AmazonSSMRoleForInstancesQuickSetup",
```

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+            let ec2Instances = describeInstances.data.map((reservation) => reservation.Instances).flat();
+
+            if (!ec2Instances.length) {
+                helpers.addResult(results, 0, 'No EC2 instances found', region);
+                return rcb();
+            }
+
+            for (let instanceInfo of describeInstanceInformation.data) {
+                const arn = `arn:aws:ec2:${region}:${accountId}:instance/${instanceInfo.InstanceId}`;
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:61]: Change aws partition

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            }
+
+            let ec2Instances = describeInstances.data.map((reservation) => reservation.Instances).flat();
+
+            if (!ec2Instances.length) {
+                helpers.addResult(results, 0, 'No EC2 instances found', region);
+                return rcb();
+            }
+
+            for (let instanceInfo of describeInstanceInformation.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:57]: Loop ec2 instances instead

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            }
+
+            var describeInstanceInformation = helpers.addSource(cache, source,
+                ['ssm', 'describeInstanceInformation', region]);
+
+            if (!describeInstanceInformation || describeInstanceInformation.err || !describeInstanceInformation.data || !describeInstanceInformation.data.length) {
+                helpers.addResult(results, 2,
+                    'No EC2 instance is being managed by AWS SSM: ' + helpers.addError(describeInstanceInformation), region);
+                return rcb();
+            }
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:48]:
```suggestion
            if (!describeInstanceInformation || describeInstanceInformation.err || !describeInstanceInformation.data) {
                helpers.addResult(results, 3,
                    'Unable to query instance information: ' + helpers.addError(describeInstanceInformation), region);
                return rcb();
            }
```

File: plugins/aws/ssm/ssmManagedInstances.spec.js [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        });
+
+        it('should FAIL if there are ec2 instances but they are not managed by AWS SSM', function (done) {
+            const cache = createCache(describeInstances, []);
+            ssmManagedInstances.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                done();
+            });
+        });
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.spec.js:243]:
```suggestion
```

File: plugins/aws/ssm/ssmManagedInstances.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            if (!describeInstanceInformation || describeInstanceInformation.err || !describeInstanceInformation.data) {
+                helpers.addResult(results, 3,
+                    'Unable to query instance information: ' + helpers.addError(describeInstanceInformation), region);
+                return rcb();
+            }
+
+            for (let ec2Instance of ec2Instances) {
+                const arn = `arn:${awsOrGov}:ec2:${region}:${accountId}:instance/${ec2Instance.InstanceId}`;
+
+                let instanceInfo = describeInstanceInformation.data.find((instanceInfo) => instanceInfo.InstanceId === ec2Instance.InstanceId);
```
Comment (AkhtarAmir) [inline plugins/aws/ssm/ssmManagedInstances.js:53]:
```suggestion
                let instanceInfo = describeInstanceInformation.data.find((instanceInfo) => instanceInfo.InstanceId && instanceInfo.InstanceId === ec2Instance.InstanceId);
```

**Review by AkhtarAmir — APPROVED** [review]

**Review by dipsubha06 — APPROVED** [review]

**Pull request merged**